### DISCUSSION TITLE
Passing variables to Pango markup language

### DISCUSSION THREAD

**Discussion by hellocatfood** [ImageMagick/ImageMagick#5035] [Help] [answered]
I'm trying to create an animation showcasing variable fonts in action. To do this I intend to render images to text using variable fonts (below using [Golos Text VF](https://v-fonts.com/fonts/golos-vf)) at various weights and then compile them into an animation using ImageMagick or ffmpeg. However in the below script I'm getting an error because Pango doesn't seem to allow variables in its markup language
```
weight=10 ; while [ $weight -le 1000 ] ; do convert -background lightblue -fill blue  -font "Golos Text VF" -pointsize 72 pango:'<span weight="$weight">"$weight" weight</span>' "$weight"_weight.png ; $weight=$(($weight + 10)) ; done
```

Is there any way to pass variables to pango markup language?

I'm using ImageMagick 6.9.11-60 Q16 x86_64 2021-01-25 on Ubuntu 21.10

**Comment by fmw42** [comment]
Do you weight computation outside of the convert command and then just pass the resulting variable?  Variable computation within commands is only possible using %[fx:...] in ImageMagick 7.

**Comment by snibgo** [comment, marked as answer]
This isn't a pango problem, but a bug in your bash script.

In bash, when a variable such as $weight is enclosed in single quotes '... $weight ...' bash does not expand the variable. So IM sees "$weight" instead of a number "10".

EDIT: You have another bug, in the final assignment. Try this:
```
$ weight=10 ; while [ $weight -le 50 ] ; do echo $weight ; $weight=$(($weight + 10)) ; done
```
It doesn't work. You need "weight=$(($weight + 10))"

**Comment by hellocatfood** [comment]
@snibgo you're right, there were errors with my script! Correct version below

`weight=10 ; while [ $weights -le 1000 ] ; do convert -background lightblue -fill blue  -font "Golos Text VF" -pointsize 72 pango:"<span weight='$weights'>"$weights" weight</span>" "$weights"_weight.png ; weights=$(($weights + 10)) ; done`

Here's an example animation
![gfasd](https://user-images.githubusercontent.com/151947/162589174-18e9f9ae-2f02-4fd8-8292-7183983226aa.gif)

**Reply by fmw42** [reply]
You still have $weights enclosed in single quotes in the span. Either remove them or replace with double quotes. Not sure why you have to have any quotes there since your span in enclosed in double quotes.

Use curly brackets {...} to ensure your variables are just for weights and not any other attached text.

try

`weight=10 ; while [ $weights -le 1000 ] ; do convert -background lightblue -fill blue -font "Golos Text VF" -pointsize 72 pango:"<span weight=${weights}>"${weights} "weight</span>" ${weights}_weight.png ; weights=$(($weights + 10)) ; done`

or perhaps just

`weight=10 ; while [ $weights -le 1000 ] ; do convert -background lightblue -fill blue -font "Golos Text VF" -pointsize 72 pango:"<span weight=${weights}>${weights} weight</span>" ${weights}_weight.png ; weights=$(($weights + 10)) ; done`

**Reply by hellocatfood** [reply]
Thanks again for this. I haven't used curly brackets in my scripts before so will have to read up on how to best use them in the future

**Reply by hellocatfood** [reply]
So I tried you edited scripts and on IM 6.9 I get this error
```
pango_layout_set_markup_with_accel: Error on line 1 char 22: Odd character “1”, expected an open quote mark after the equals sign when giving value for attribute “weight” of element “span”
convert-im6.q16: Error on line 1 char 22: Odd character “1”, expected an open quote mark after the equals sign when giving value for attribute “weight” of element “span” `<span weight=160>160 weight</span>' @ error/pango.c/ReadPANGOImage/362.
convert-im6.q16: negative or zero image size `<span weight=160>160 weight</span>' @ error/image.c/SetImageExtent/2688.
convert-im6.q16: no images defined `160_weight.png' @ error/convert.c/ConvertImageCommand/3229.
Aborted (core dumped)
```

and on 7.0 I get this error
```
magick: unable to open image '<span weight=10>10': No such file or directory @ error/blob.c/OpenBlob/3527.
```

**Comment by snibgo** [comment]
> You still have $weights enclosed in single quotes in the span. Either remove them or replace with double quotes. Not sure why you have to have any quotes there since your span in enclosed in double quotes.

Quotes are required. Within the Pango markup, attributes should be quoted. See examples at https://legacy.imagemagick.org/Usage/text/#pango
